subprojects/emcee/worker/src/main/kotlin/com/avito/emcee/worker/internal/rest/HttpServer.kt

```diff
@@ -1,93 +1,55 @@
 package com.avito.emcee.worker.internal.rest
 
 import com.avito.emcee.worker.internal.rest.handler.RequestHandler
-import java.io.BufferedReader
-import java.io.InputStreamReader
-import java.io.PrintWriter
-import java.net.ServerSocket
-import java.net.Socket
+import io.ktor.serialization.gson.gson
+import io.ktor.server.application.call
+import io.ktor.server.application.install
+import io.ktor.server.engine.embeddedServer
+import io.ktor.server.netty.Netty
+import io.ktor.server.netty.NettyApplicationEngine
+import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.server.response.respond
+import io.ktor.server.routing.route
+import io.ktor.server.routing.routing
 
 internal class HttpServer(
-    private val handlers: List<RequestHandler>,
+    private val handlers: List<RequestHandler<*>>,
     private val debug: Boolean
 ) {
 
-    private var serverSocket: ServerSocket? = null
-
-    @Volatile
-    internal var isStarted: Boolean = false
+    private var server: NettyApplicationEngine? = null
 
     fun start(port: Int) {
         if (debug) println("Starting REST server on $port port/")
-        serverSocket = ServerSocket(port)
-        isStarted = true
-        while (true) {
-            ClientHandler(
-                clientSocket = serverSocket!!.accept(),
-                debug = debug,
-                handlers = handlers
-            ).start()
-        }
-    }
-
-    fun stop() {
-        serverSocket?.close()
-        isStarted = false
-    }
-
-    @Suppress("IfThenToElvis")
-    private class ClientHandler(
-        private val clientSocket: Socket,
-        private val debug: Boolean,
-        private val handlers: List<RequestHandler>
-    ) : Thread() {
-
-        override fun run() {
-            val reader = BufferedReader(InputStreamReader(clientSocket.getInputStream()))
-            val writer = PrintWriter(clientSocket.getOutputStream())
-
-            try {
-                val request = reader.readLine() ?: return
-
-                val body = StringBuilder()
-                while (true) {
-                    val line = reader.readLine()
-                    if (line == null || line.isEmpty()) break
-                    body.append(line).append("\n")
-                }
-
-                if (debug) println("<-- $request\n$body")
-
-                val handler = findHandler(request)
-                val response = if (handler != null) {
-                    handler.response().wrapToSuccessfulResponse()
-                } else {
-                    badRequest()
+        server = embeddedServer(Netty, port = port) {
+            install(ContentNegotiation) {
+                gson() // TODO: use Moshi instead when it will be available
+            }
+            routing {
+                handlers.forEach { handler: RequestHandler<*> ->
+                    route(handler.path, handler.method) {
+                        handle {
+                            call.respond(handler.response())
```

Comment:
Could we describe routes explicitly there without RequestHandler abstraction?

Reply:
Yes, we can. But then we'd have to provide all necessary dependencies to the `HttpServer` directly. It brakes the dependency inversion principle and increases the coupling degree.

Reply:
What do you think about declaring multiple methods in Retrofit service? Does that break DI principle?
As for me I want to see REST API of one service in one place

Reply:
We could extract routing describing in separate file